src/qt/optionsmodel.cpp

```diff
@@ -19,18 +19,103 @@
 
 #include <QDebug>
 #include <QLatin1Char>
 #include <QSettings>
 #include <QStringList>
 
+#include <univalue.h>
+
 const char *DEFAULT_GUI_PROXY_HOST = "127.0.0.1";
 
 static const QString GetDefaultProxyAddress();
 
-OptionsModel::OptionsModel(QObject *parent, bool resetSettings) :
-    QAbstractListModel(parent)
+//! Convert QVariant value to bitcoin setting of specified type. Type is
+//! explicitly required to make sure values written to settings.json are always
+//! written with the right type, regardless of how the options dialog code sets
+//! the QVariant.
+static util::SettingsValue ToSetting(const QVariant& variant, QVariant::Type type)
+{
+    assert(type == QVariant::Bool || type == QVariant::Int || type == QVariant::String);
+    if (!variant.isValid()) return {};
+    if (type == QVariant::Bool) return variant.toBool();
+    if (type == QVariant::Int) return variant.toInt();
+    return variant.toString().toStdString();
+}
+
+//! Convert bitcoin setting to QVariant. This is used to interpret values from
+//! settings.json, which have JSON types, and values from bitcoin.conf, which
+//! are untyped strings. The return value is not guaranteed to have any
+//! particular type, so if a specific type is required, ToQString() or ToInt()
+//! methods below should be used instead of this.
+static QVariant ToQVariant(const util::SettingsValue& value, const QVariant& fallback = {})
+{
+    if (value.isNull()) return fallback;
+    if (value.isBool()) return value.get_bool();
+    if (value.isNum()) return value.get_int();
+    return QString::fromStdString(value.get_str());
+}
+
+//! Convert bitcoin setting to QString. The setting must already be a string, or
+//! it must be negated or unset. If it has another type like integer, bool,
+//! array, or object, this will raise an exception.
+static QString ToQString(const util::SettingsValue& value, const QString& fallback = {})
+{
+    if (value.isNull()) return fallback; // If string setting is unset, fall back to default.
+    if (value.isFalse()) return "";      // If string setting is negated, treat it like "".
+    return QString::fromStdString(value.get_str());
+}
+
+//! Convert bitcoin setting to integer. The setting must already be an integer
+//! or integer string, or it must be negated or unset. If it has another type,
+//! this will raise an exception.
+static int ToInt(const util::SettingsValue& value, int fallback = 0)
+{
+    if (value.isNull()) return fallback; // If int setting is unset, fall back to default.
+    if (value.isFalse()) return 0;       // If int setting is negated, treat it like 0.
+    if (value.isNum()) return value.get_int();
+    return LocaleIndependentAtoi<int>(value.get_str());
+}
+
+//! Get pruning enabled value to show in GUI from bitcoin -prune setting.
+static bool PruneEnabled(const util::SettingsValue& prune_setting)
+{
+    // -prune=1 setting is manual pruning mode, so disabled for purposes of the gui
+    return ToInt(prune_setting) > 1;
```

Comment:
This, in isolation, is (wrongly) ignoring the fact that values in [2, 550) are not allowed (`MIN_DISK_SPACE_FOR_BLOCK_FILES` -- 550 MiB). `init.cpp` would refuse to start with such a value.

In context, this is only called with values entered in the GUI which is accepting integer GB only, so it is impossible to enter e.g. 100 MiB. Or it could come from `settings.json` where the value is stored in MiB and the user may have edited it (they are not supposed to). I think it is ok because even if <550 is entered in `settings.json` the startup would gracefully fail with the proper message from `init.cpp`.

Reply:
Can be marked as resolved.